drivers/dma/dma_stm32u5.c

```diff
@@ -0,0 +1,832 @@
+{
+	LL_DMA_ConfigAddresses(dma, channel, src_addr, dest_addr);
+}
+
+/* same function to set periph/mem addresses */
+void stm32_dma_set_periph_mem_address(DMA_TypeDef *dma,
+					     uint32_t channel,
+					     uint32_t src_addr,
+					     uint32_t dest_addr)
+{
+	LL_DMA_ConfigAddresses(dma, channel, src_addr, dest_addr);
+}
+
+static void dma_stm32_irq_handler(const struct device *dev, uint32_t id)
+{
+	const struct dma_stm32_config *config = dev->config;
+	DMA_TypeDef *dma = (DMA_TypeDef *)(config->base);
+	struct dma_stm32_stream *stream;
+	uint32_t callback_arg;
+
+	__ASSERT_NO_MSG(id < config->max_streams);
+
+	stream = &config->streams[id];
+	/* Exit if stream is no more busy */
+	if (stream->busy == false) {
+		dma_stm32_clear_stream_irq(dev, id);
+		return;
+	}
+	callback_arg = id + STM32_DMA_STREAM_OFFSET;
+	stream->busy = false;
+
+	/* The dma stream id is in range from STM32_DMA_STREAM_OFFSET..<dma-requests> */
+	if (stm32_dma_is_ht_irq_active(dma, id)) {
+		/* Let HAL DMA handle flags on its own */
+		if (!stream->hal_override) {
+			dma_stm32_clear_ht(dma, id);
+		}
+		stream->dma_callback(dev, stream->user_data, callback_arg, 0);
+	} else if (stm32_dma_is_tc_irq_active(dma, id)) {
+#ifdef CONFIG_DMAMUX_STM32
+		stream->busy = false;
+#endif
+		/* Let HAL DMA handle flags on its own */
+		if (!stream->hal_override) {
+			dma_stm32_clear_tc(dma, id);
+		}
+		stream->dma_callback(dev, stream->user_data, callback_arg, 0);
+	} else if (stm32_dma_is_unexpected_irq_happened(dma, id)) {
+		LOG_ERR("Unexpected irq happened.");
+		stream->dma_callback(dev, stream->user_data,
+				     callback_arg, -EIO);
+	} else {
+		LOG_ERR("Transfer Error.");
+		dma_stm32_dump_stream_irq(dev, id);
+		dma_stm32_clear_stream_irq(dev, id);
+		stream->dma_callback(dev, stream->user_data,
+				     callback_arg, -EIO);
+	}
+}
+
+static int dma_stm32_get_priority(uint8_t priority, uint32_t *ll_priority)
+{
+	if (priority > ARRAY_SIZE(table_priority)) {
+		LOG_ERR("Priority error. %d", priority);
+		return -EINVAL;
+	}
+
+	*ll_priority = table_priority[priority];
+	return 0;
+}
+
+static int dma_stm32_get_direction(enum dma_channel_direction direction,
+				   uint32_t *ll_direction)
+{
+	switch (direction) {
+	case MEMORY_TO_MEMORY:
+		*ll_direction = LL_DMA_DIRECTION_MEMORY_TO_MEMORY;
+		break;
+	case MEMORY_TO_PERIPHERAL:
+		*ll_direction = LL_DMA_DIRECTION_MEMORY_TO_PERIPH;
+		break;
+	case PERIPHERAL_TO_MEMORY:
+		*ll_direction = LL_DMA_DIRECTION_PERIPH_TO_MEMORY;
+		break;
+	default:
+		LOG_ERR("Direction error. %d", direction);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int dma_stm32_disable_stream(DMA_TypeDef *dma, uint32_t id)
+{
+	int count = 0;
+
+	for (;;) {
+		if (stm32_dma_disable_stream(dma, id) == 0) {
+			return 0;
+		}
+		/* After trying for 5 seconds, give up */
+		if (count++ > (5 * 1000)) {
+			return -EBUSY;
+		}
+		k_sleep(K_MSEC(1));
+	}
+
+	return 0;
+}
+
+DMA_STM32_EXPORT_API int dma_stm32_configure(const struct device *dev,
```

Comment:
Is DMA_STM32_EXPORT_API required here ?

Reply:
Since the STM32U5 has _no_ DMAMUX for GPDMA, the DMA_STM32_EXPORT_API is known as 'static'